### PR TITLE
fix(gomod): Adjust regex used to disable replace directives

### LINKED ISSUE #11428: A `replace` directive in a `go.mod` won't be ignored if it's a multi-line block (while single-liners are ignored as expected)
### How are you running Renovate?

Self-hosted

### Please select which platform you are using if self-hosting.

Bitbucket Cloud (bitbucket.org)

### If you're self-hosting Renovate, tell us what version of Renovate you run.

v26.6.0

### Describe the bug

(Probably related to https://github.com/renovatebot/renovate/issues/6647)

A `replace` directive in a `go.mod` file can either be a single-liner or span multiple lines. Both of the following are valid `replace` directives. (See https://golang.org/ref/mod#go-mod-file-replace for further information.)
```
replace example.com/module_a => ../module_a

replace (
    example.com/module_a => ../module_a
    example.com/module_b => ../module_b
)
```

However, it seems that Renovate is only capable of handling the single-line case, as is suggested by the following code:
https://github.com/renovatebot/renovate/blob/d260dbf2d970b2d097b45b59cb3067ac532916a7/lib/manager/gomod/artifacts.ts#L109-L112

In the multi-line case Renovate will fail to run `go get -d ./...` because the `replace`-ments are not being ignored.

I've created the following repositories (on GitHub though) to demonstrate the issue.
- https://github.com/renovate-gomod-issue/foo, a module having tags v1.0.0 and v1.0.1.
- https://github.com/renovate-gomod-issue/bar, another module having tags v1.0.0 and v1.0.1.
- https://github.com/renovate-gomod-issue/partially-replaced, a module depending on `foo` v1.0.0 and `bar` v1.0.0, where `foo` is replaced.
- https://github.com/renovate-gomod-issue/fully-replaced, another module depending on `foo` v1.0.0 and `bar` v1.0.0, where both `foo` and `bar` are replaced.

I've [forked and edited Renovate](https://github.com/renovate-gomod-issue/renovate/pull/1/files#diff-b53c48291df312e71cae3acd27c0f571c2d6c0234a4f2444e0293196a4d76f05) so that it prints out the content of any `go.mod` before and after the "massaging".

When I ran `RENOVATE_TOKEN=xxx npm start` with [the config here](https://github.com/renovate-gomod-issue/renovate/pull/1/files#diff-e8e9eea3276b2de9a74fd257902dc74c846f6fc8653bbddf45d738b71201f93d), Renovate succeeded with the `partially-replaced` repository and failed with the `fully-replaced` repository.

The complete logs are included in the branch as well:
- [stdout](https://github.com/renovate-gomod-issue/renovate/pull/1/files#diff-275f8eb0b3670719242ead5792bf63d698bda7b3a0506a3a7d2880a8b44fecb9)
- [renovate.log](https://github.com/renovate-gomod-issue/renovate/pull/1/files#diff-17ca9073197e56ebb6828c2911e7aeb983bc6c17ec780ea4cf6305df3ec1545c)

### Relevant debug logs

<details><summary>Logs</summary>

```
 INFO: Lock file error - aborting (repository=renovate-gomod-issue/fully-replaced)

{"name":"renovate","hostname":"localhost.localdomain","pid":479336,"level":20,"logContext":"NcssAJ-sw","repository":"renovate-gomod-issue/fully-replaced","branch":"renovate/github.com-renovate-gomod-issue-bar-1.x","msg":"gomod.updateArtifacts(go.mod)","time":"2021-08-25T14:22:11.336Z","v":0}
{"name":"renovate","hostname":"localhost.localdomain","pid":479336,"level":20,"logContext":"NcssAJ-sw","repository":"renovate-gomod-issue/fully-replaced","branch":"renovate/github.com-renovate-gomod-issue-bar-1.x","msg":"Compare go.mod before and after massaging","time":"2021-08-25T14:22:11.337Z","v":0}
{"name":"renovate","hostname":"localhost.localdomain","pid":479336,"level":20,"logContext":"NcssAJ-sw","repository":"renovate-gomod-issue/fully-replaced","branch":"renovate/github.com-renovate-gomod-issue-bar-1.x","msg":"module github.com/renovate-gomod-issue/fully-replaced\n\ngo 1.16\n\nrequire (\n\tgithub.com/renovate-gomod-issue/bar v1.0.1\n\tgithub.com/renovate-gomod-issue/foo v1.0.0\n)\n\nreplace (\n\tgithub.com/renovate-gomod-issue/bar => ../bar\n\tgithub.com/renovate-gomod-issue/foo => ../foo\n)\n","time":"2021-08-25T14:22:11.337Z","v":0}
{"name":"renovate","hostname":"localhost.localdomain","pid":479336,"level":20,"logContext":"NcssAJ-sw","repository":"renovate-gomod-issue/fully-replaced","branch":"renovate/github.com-renovate-gomod-issue-bar-1.x","msg":"module github.com/renovate-gomod-issue/fully-replaced\n\ngo 1.16\n\nrequire (\n\tgithub.com/renovate-gomod-issue/bar v1.0.1\n\tgithub.com/renovate-gomod-issue/foo v1.0.0\n)\n\nreplace (\n\tgithub.com/renovate-gomod-issue/bar => ../bar\n\tgithub.com/renovate-gomod-issue/foo => ../foo\n)\n","time":"2021-08-25T14:22:11.337Z","v":0}
{"name":"renovate","hostname":"localhost.localdomain","pid":479336,"level":20,"logContext":"NcssAJ-sw","repository":"renovate-gomod-issue/fully-replaced","branch":"renovate/github.com-renovate-gomod-issue-bar-1.x","cmd":"go","args":"get -d ./...","msg":"go get command included","time":"2021-08-25T14:22:11.342Z","v":0}
{"name":"renovate","hostname":"localhost.localdomain","pid":479336,"level":20,"logContext":"NcssAJ-sw","repository":"renovate-gomod-issue/fully-replaced","branch":"renovate/github.com-renovate-gomod-issue-bar-1.x","command":["go get -d ./..."],"msg":"Executing command","time":"2021-08-25T14:22:11.343Z","v":0}
{"name":"renovate","hostname":"localhost.localdomain","pid":479336,"level":20,"logContext":"NcssAJ-sw","repository":"renovate-gomod-issue/fully-replaced","branch":"renovate/github.com-renovate-gomod-issue-bar-1.x","err":{"killed":false,"code":1,"signal":null,"cmd":"go get -d ./...","stdout":"","stderr":"go: github.com/renovate-gomod-issue/bar@v1.0.1 (replaced by ../bar): reading ../bar/go.mod: open /tmp/renovate/repos/github/renovate-gomod-issue/bar/go.mod: no such file or directory\n","message":"Command failed: go get -d ./...\ngo: github.com/renovate-gomod-issue/bar@v1.0.1 (replaced by ../bar): reading ../bar/go.mod: open /tmp/renovate/repos/github/renovate-gomod-issue/bar/go.mod: no such file or directory\n","stack":"Error: Command failed: go get -d ./...\ngo: github.com/renovate-gomod-issue/bar@v1.0.1 (replaced by ../bar): reading ../bar/go.mod: open /tmp/renovate/repos/github/renovate-gomod-issue/bar/go.mod: no such file or directory\n\n    at ChildProcess.exithandler (child_process.js:319:12)\n    at ChildProcess.emit (events.js:376:20)\n    at ChildProcess.emit (domain.js:470:12)\n    at maybeClose (internal/child_process.js:1055:16)\n    at Socket.<anonymous> (internal/child_process.js:441:11)\n    at Socket.emit (events.js:376:20)\n    at Socket.emit (domain.js:470:12)\n    at Pipe.<anonymous> (net.js:673:12)"},"msg":"rawExec err","time":"2021-08-25T14:22:11.370Z","v":0}
{"name":"renovate","hostname":"localhost.localdomain","pid":479336,"level":20,"logContext":"NcssAJ-sw","repository":"renovate-gomod-issue/fully-replaced","branch":"renovate/github.com-renovate-gomod-issue-bar-1.x","err":{"killed":false,"code":1,"signal":null,"cmd":"go get -d ./...","stdout":"","stderr":"go: github.com/renovate-gomod-issue/bar@v1.0.1 (replaced by ../bar): reading ../bar/go.mod: open /tmp/renovate/repos/github/renovate-gomod-issue/bar/go.mod: no such file or directory\n","message":"Command failed: go get -d ./...\ngo: github.com/renovate-gomod-issue/bar@v1.0.1 (replaced by ../bar): reading ../bar/go.mod: open /tmp/renovate/repos/github/renovate-gomod-issue/bar/go.mod: no such file or directory\n","stack":"Error: Command failed: go get -d ./...\ngo: github.com/renovate-gomod-issue/bar@v1.0.1 (replaced by ../bar): reading ../bar/go.mod: open /tmp/renovate/repos/github/renovate-gomod-issue/bar/go.mod: no such file or directory\n\n    at ChildProcess.exithandler (child_process.js:319:12)\n    at ChildProcess.emit (events.js:376:20)\n    at ChildProcess.emit (domain.js:470:12)\n    at maybeClose (internal/child_process.js:1055:16)\n    at Socket.<anonymous> (internal/child_process.js:441:11)\n    at Socket.emit (events.js:376:20)\n    at Socket.emit (domain.js:470:12)\n    at Pipe.<anonymous> (net.js:673:12)"},"msg":"Failed to update go.sum","time":"2021-08-25T14:22:11.370Z","v":0}
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by tfussell**
## Changes

This PR splits the code that disables `replace` directives into a separate function to make it more easily tested (without all of the setup and teardown used by the other tests). Additionally, it changes the disabling logic by modifying the regex so that only relative replace directives are commented-out to fix a problem with Renovate PRs having a dirty `go mod tidy` for repos containing replace directives (normal ones, not relative paths).

## Context

In https://github.com/renovatebot/renovate/pull/14033, multi-line replace directives were disabled entirely to fix https://github.com/renovatebot/renovate/issues/11428. Looking through GitHub history, I found in https://github.com/renovatebot/renovate/issues/2596 that this logic (implemented in https://github.com/renovatebot/renovate/commit/9551ed75c0c1e4cc0401f4d82141db43e67d84d8) was originally only intended to disable replace directives that point to a parent directory (i.e. starts with `../`) so I believe the new regexes in #14033 are overly broad.

To explain my use case, my company uses Renovate on many repos. We also have CI checks which ensure that `go mod tidy` shows no changes. Due to the potential bug described in the previous paragraph, almost all Renovate PRs opened on Go repos containing any replace directive fail CI and therefore aren't working for us. I can provide some diffs to demonstrate the result of Renovate versus running `go mod tidy` locally or in CI. I have already verified that it's not a version mismatch using the same version of Go as Renovate.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/gomod/artifacts.ts [hunk 1]
```diff
@@ -156,37 +165,7 @@ export async function updateArtifacts({
...
-     */
-    const blockCommentOut = (match): string =>
-      match.replace(/(\r?\n)/g, '$1// renovate-replace ');
-
-    // Comment out golang replace directives
-    const massagedGoMod = newGoModContent
-      .replace(inlineReplaceRegEx, inlineCommentOut)
-      .replace(blockReplaceRegEx, blockCommentOut);
-
+    const massagedGoMod = disableLocalParentReplaceDirective(newGoModContent);
```
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.ts:168]: This looks like a refactor and a fix PR, that should be done in separate PR for easier review.

File: lib/modules/manager/gomod/artifacts.spec.ts [hunk 2]
```diff
@@ -28,10 +28,65 @@ require golang.org/x/foo v1.0.0
 require github.com/rarkins/foo abcdef1
 require gopkg.in/russross/blackfriday.v1 v1.0.0
 
+replace gopkg.in/russross/blackfriday.v1 ./blackfriday
+
 replace github.com/pkg/errors => ../errors
 
-replace (golang.org/x/foo => github.com/pravesht/gocql v0.0.0)
```
[the remark this answers is not in the document]
Comment (tfussell) [reply lib/modules/manager/gomod/artifacts.spec.ts:33]: That's not a valid replace block as far as I can tell. If I try it locally, `go mod tidy` errors with
```
usage: replace module/path [v1.2.3] => other/module v1.4
	 or replace module/path [v1.2.3] => ../local/directory
```
According to https://go.dev/ref/mod#go-mod-file-replace, the opening and closing parentheses and each replacement statement in the block have to be followed by a newline.

Also it is possible to have mixed replace statements, single line and block, in a single `go.mod`, but that's very rare in my experience.

File: lib/modules/manager/gomod/artifacts.ts (hunk 1 above)
Comment (tfussell) [reply lib/modules/manager/gomod/artifacts.ts:168]: I was primarily interested in getting the bug fixed, but I didn't see an obvious way to test the replacement logic in isolation without splitting it into a separate function. If you have any resources on testing this using integration tests or on real repositories, I can revert the unit tests and test this PR that way.

File: lib/modules/manager/gomod/artifacts.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/manager/gomod/artifacts.spec.ts:33]: Ok, i want to make sure this also works on older go versions, where that maybe was allowed. 🤷‍♂️

[the remark this answers is not in the document]
Comment (tfussell) [reply lib/modules/manager/gomod/artifacts.spec.ts:33]: I just tried it with go 1.13, the first version where modules were the default, and got the same error.

[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/manager/gomod/artifacts.spec.ts:33]: Thanks for verifying.

**Comment by fiunchinho** [comment]
How can we move this forward?

**Comment by rarkins** [comment]
> How can we move this forward?

This PR needs to be updated according to the maintainer's request.

I'll also add that it has removed some well commented code and replaced it with an undocumented regex, which is undesirable too.
